### PR TITLE
privilege: add `authHost` and use `authHost` to verificate

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
TiDB will fail in this case:
```
// mysql -u root -p -h 192.168.198.7 -P 4000
mysql> create user 'root'@'192.168.%';
Query OK, 0 rows affected (0.01 sec)

mysql> select current_user();
+----------------+
| current_user() |
+----------------+
| root@%         |
+----------------+
1 row in set (0.00 sec)

mysql> grant all on *.* to 'root'@'192.168.%';
ERROR 1105 (HY000): privilege check fail
mysql>
```

When `'root'@'192.168.%'` was created, although current user is still `'root'@'%'`, host field in `UserManager` is `192.168.198.7` which will match with `192.168.%`. TiDB do verification use `HostName` rather than `AuthHostName`.

### What is changed and how it works?
Add `authHost` field which was init at logging in. For example, if `'root'@'%'` log in  at `192.168.0.10`, it's `authHost` is `%`. TiDB always use this `authHost` to  verificate.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: privilege/privileges/privileges.go [hunk 1]
```diff
@@ -139,6 +140,7 @@ func (p *UserPrivileges) ConnectionVerification(user, host string, authenticatio
 	if len(pwd) == 0 && len(authentication) == 0 {
 		p.user = user
 		p.host = host
+		p.authHost = h
```
Comment (tiancaiamao) [inline privilege/privileges/privileges.go]: Should we directly set `p.user = u` and `p.host = h`?
Where is `p.host` used after this change?

Comment (imtbkcat) [reply privilege/privileges/privileges.go]: fixed, PTAL @tiancaiamao

File: privilege/privileges/privileges.go [hunk 2]
```diff
@@ -158,7 +158,7 @@ func (p *UserPrivileges) ConnectionVerification(user, host string, authenticatio
 	}
 
 	p.user = user
```
Comment (tiancaiamao) [inline privilege/privileges/privileges.go]: Should it be `p.user = u` here?

I'm not sure could the privilege record be something like 
user = '%' and host = '192.168.%'

**Comment by tiancaiamao** [comment]
`make fmt` fail. @imtbkcat

File: privilege/privileges/privileges.go (hunk 2 above)
Comment (imtbkcat) [reply privilege/privileges/privileges.go]: @tiancaiamao MySQL dosen't support wildcard in user name.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**